Fix Launch4j's classpath if "dontWrapJar" is "true", add tests

sschuberth opened the pull request:
According to [1] the Java plugin's default output directory is "libs", not
"lib". Fix this by generalizing the way we get the output directory.

[1] https://docs.gradle.org/current/userguide/java_plugin.html#N1245A

TheBoegl commented:
Thanks for your pull request.
I'm looking into merging and adopting your fix though it is not compatible with launch4j right now:
`launch4j: Specify runtime path of the jar relative to the executable.`

sschuberth commented on src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy:
```diff
@@ -21,7 +21,12 @@ class CreateLaunch4jXMLTask extends DefaultTask {
+        def classpath = []
+        if (project.plugins.hasPlugin('java')) {
+            classpath = project.configurations.runtime.collect {
+                "${project.tasks[JavaPlugin.JAR_TASK_NAME].destinationDir}/${it.name}"
```
Odd, I thought I had tested this. For testing, does it help to say `project.relativePath(project.tasks[JavaPlugin.JAR_TASK_NAME].destinationDir + File.separator + it.name)` here?

TheBoegl replied:
I'll test these lines with a new test project.

sschuberth commented:
I've updated the PR to make this easier for you to check.

aaiezza commented:
Excellent work!

On Mon, Aug 10, 2015 at 8:55 AM, Sebastian Schuberth <
notifications@github.com> wrote:

> According to [1] the Java plugin's default output directory is "libs", not
> "lib". Fix this by generalizing the way we get the output directory.
>
> [1] https://docs.gradle.org/current/userguide/java_plugin.html#N1245A
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/TheBoegl/gradle-launch4j/pull/9
> Commit Summary
>
>    - Fix the default jar path
>
> File Changes
>
>    - *M* README.md
>    <https://github.com/TheBoegl/gradle-launch4j/pull/9/files#diff-0> (2)
>    - *M* src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy
>    <https://github.com/TheBoegl/gradle-launch4j/pull/9/files#diff-1> (7)
>    - *M*
>    src/main/groovy/edu/sc/seis/launch4j/Launch4jPluginExtension.groovy
>    <https://github.com/TheBoegl/gradle-launch4j/pull/9/files#diff-2> (2)
>
> Patch Links:
>
>    - https://github.com/TheBoegl/gradle-launch4j/pull/9.patch
>    - https://github.com/TheBoegl/gradle-launch4j/pull/9.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/TheBoegl/gradle-launch4j/pull/9>.
>

TheBoegl commented on src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy:
```diff
@@ -21,7 +21,12 @@ class CreateLaunch4jXMLTask extends DefaultTask {
+        def classpath = []
+        if (project.plugins.hasPlugin('java')) {
+            classpath = project.configurations.runtime.collect {
+                project.relativePath(project.tasks[JavaPlugin.JAR_TASK_NAME].destinationDir + File.separator + it.name)
```
This will not run under windows.
The correct way would be to either hardcode it (`project.tasks[JavaPlugin.JAR_TASK_NAME].destinationDir + '/' + it.name`) or use something like `project.relativePath(new File("${project.tasks[JavaPlugin.JAR_TASK_NAME].destinationDir}", it.name))`

TheBoegl commented:
@sschuberth Thanks for your contribution, but your provided patch is not working:

1. The changed class path results in a non functioning executable ('Error: A JNI error has occurred, please check your installation and try again'). To see this set `dontWrapJar` to false or don't set it at all and run the created executable.
1. The `launch4j.jar` variable is used in the launch4j way instead the gradle one: Take a look at the <jar> section in [the docs](http://launch4j.sourceforge.net/docs.html#Configuration_file)

> Optional, by default specifies the jar to wrap. To launch a jar without wrapping it enter the runtime path of the jar relative to the executable and set `<dontWrapJar>` to true.

It seems #11 fooled me and you, too

@aaiezza Does this really work for you or was the executable not created again (see #11)?

sschuberth commented:
I think it would be good to have some tests and integration with CI like AppVeyor to get deterministic results. I'll see what I can do.

sschuberth commented:
Now that #14 adds tests automation, my plan is to add more tests that discover breakages as introduced by this PR, and then update this PR with a solution that actually works without breaking anything.

TheBoegl commented:
Thanks for your contribution! Adding more test would be great.

sschuberth commented:
I've pushed an updated version that adds a test for further discussion.

@TheBoegl Could you elaborate on the differences between the `createExe`, `createExeWithBin`, `createExeWithJar` and `launch4j` tasks? What do you mean by "placeholder task" in the [README](https://github.com/TheBoegl/gradle-launch4j/blob/master/README.md)? Do you mean "alias task" instead? Which is the task I'm supported to run when I want to create a self-contained executable?

sschuberth commented:
> To see this set dontWrapJar to false or don't set it at all and run the created executable.

I've added another test which does not set `dontWrapJar` and runs the created executable. For me the test passes just fine.

TheBoegl commented:
@sschuberth The `createExe` task was the previous task name and is kept for backward compatibility. Therefore it *calls* - in gradle terms: *depends* - on the other tasks `createExeWithJar` and `createExeWithBin`. The latter is just the old renamed `createExe` task and calls the launch4j binary anywhere on the `$PATH`. The property `externalLaunch4j` decides which of the two tasks is called.
I've kept the `lauch4j` task for backward compatibility, too, though it's only  purpose is to call all other tasks.
If you would like to create an executable you were supposed to run one of the `launch4j` or `createExe` tasks and that has not changed. Anyway I'm inclined to the view that the `launch4j` task is and will be the one to call.
I should announce and clarify that in the [README](https://github.com/TheBoegl/gradle-launch4j/blob/master/README.md).
I'll do that and check your PR within the next few days.

TheBoegl commented:
Please apply the attached patches to see it failing:
```
From 09069fe4d4bd55aadf36288336d23d6e94ab5fc3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20B=C3=B6gl?= <info@sebastianboegl.de>
Date: Sun, 23 Aug 2015 21:36:40 +0200
Subject: [PATCH 1/2] add missing import and fix wrong groovy File usage

---
 src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy b/src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy
index 01bba88..1fdf314 100644
--- a/src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy
+++ b/src/main/groovy/edu/sc/seis/launch4j/CreateLaunch4jXMLTask.groovy
@@ -2,6 +2,7 @@ package edu.sc.seis.launch4j
 
 import groovy.xml.MarkupBuilder
 import org.gradle.api.DefaultTask
+import org.gradle.api.plugins.JavaPlugin
 import org.gradle.api.tasks.OutputFile
 import org.gradle.api.tasks.TaskAction
 import org.slf4j.Logger
@@ -24,7 +25,7 @@ class CreateLaunch4jXMLTask extends DefaultTask {
         def classpath = []
         if (project.plugins.hasPlugin('java')) {
             classpath = project.configurations.runtime.collect {
-                project.relativePath(project.tasks[JavaPlugin.JAR_TASK_NAME].destinationDir + File.separator + it.name)
+                project.relativePath(new File(project.tasks[JavaPlugin.JAR_TASK_NAME].destinationDir, it.name))
             }
         }
         def file = getXmlOutFile()
-- 
2.5.0
```
```
From dc0f47e52f0a6a0588843c298e8e181b3f9fb9ce Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20B=C3=B6gl?= <info@sebastianboegl.de>
Date: Sun, 23 Aug 2015 21:37:35 +0200
Subject: [PATCH 2/2] add dependencie for test project

---
 .../edu/sc/seis/launch4j/Launch4jPluginExtensionTest.groovy  | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/test/groovy/edu/sc/seis/launch4j/Launch4jPluginExtensionTest.groovy b/src/test/groovy/edu/sc/seis/launch4j/Launch4jPluginExtensionTest.groovy
index 1ec1b78..9b8afae 100644
--- a/src/test/groovy/edu/sc/seis/launch4j/Launch4jPluginExtensionTest.groovy
+++ b/src/test/groovy/edu/sc/seis/launch4j/Launch4jPluginExtensionTest.groovy
@@ -105,6 +105,11 @@ class Launch4jPluginExtensionTest extends Specification {
             apply plugin: 'edu.sc.seis.launch4j'
             apply plugin: 'java'
 
+            dependencies {
+                compile 'org.slf4j:slf4j-api:1.7.12'
+                runtime 'org.slf4j:slf4j-nop:1.7.12'
+            }
+
             launch4j {
                 mainClassName = 'com.test.app.Main'
                 outfile = 'test.exe'
@@ -115,9 +120,16 @@ class Launch4jPluginExtensionTest extends Specification {
         sourceFile << """
             package com.test.app;
 
+            import org.slf4j.Logger;
+            import org.slf4j.LoggerFactory;
+
             public class Main {
+
+                private static final Logger LOG = LoggerFactory.getLogger(Main.class);
+
                 public static void main(String[] args) {
                     System.out.println("Hello World!");
+                    LOG.info("Hello World!");
                 }
             }
         """
-- 
2.5.0
```

sschuberth commented:
@TheBoegl Thanks, I've added your test (with minor modifications) and fixed the issue. I think we're good now, or?

TheBoegl commented:
@sschuberth I thought as much, though this is not the case here.
Your change in sschuberth@bcadeec62252363118d483ec1a5c6eedc3bb0e6f breaks the portability of the executable. Just start an executable created for an example project with the launch4j debugging flag `--l4j-debug` and check the line *Launcher args* or *Add classpath* or execute it on a remote machine. Mine uses the libraries from my gradle cache.
I did neither get the test cases to fail nor even the created executable from these tests *test.exe* to print to my command prompt, so I feel "kinda lost". 

If you add another test case setting the `dontWrapJar` to `true` you would have to apply this patch
```
Index: src/main/groovy/edu/sc/seis/launch4j/Launch4jPluginExtension.groovy
<+>windows-1252
===================================================================
--- src/main/groovy/edu/sc/seis/launch4j/Launch4jPluginExtension.groovy	(date 1440431833000)
+++ src/main/groovy/edu/sc/seis/launch4j/Launch4jPluginExtension.groovy	(revision )
@@ -7,6 +7,7 @@
 import org.gradle.listener.ActionBroadcast
 
 import javax.inject.Inject
+import java.nio.file.Paths
 
 class Launch4jPluginExtension implements Serializable {
 
@@ -94,7 +95,7 @@
     public String getJar(Project project) {
         if (jar == null) {
             if (project.plugins.hasPlugin('java')) {
-                jar = project.tasks[JavaPlugin.JAR_TASK_NAME].archivePath
+                jar = Paths.get(project.buildDir.path, outputDir).relativize(Paths.get(project.tasks[JavaPlugin.JAR_TASK_NAME].archivePath.path))
             } else {
                 jar = ''
             }
```
though in the end the error is the same as above.

sschuberth commented:
Turns out that I had a mayor misunderstanding WRT how JARs are wrapped. I totally overlooked that there already is a `copyL4jLib` task that copies *all* JARs the application depends on to the `lib` (not `libs`) directory. While `libs` is Gradle's standard output directory for libraries, `lib` is a custom directory used by this plugin to collect all libraries, i.e. they get copied from the Gradle cache to `lib`.

This `libs` vs. `lib` totally confused me, along with the fact the `classpath` is not set correctly if "dontWrapJar" is "true". So it never was the directory that was wrong, but the classpath. I've reworked my branch to fix that issue instead, and added one more test with `dontWrapJar = true`.

All tests pass now, please again have a look.

TheBoegl commented:
This is a fix for lazy programmers who deliver non executable jar files.
I've understood the launch4j documentation regarding the `dontWrapJar` flag different:

> Launch4j by default wraps jars in native executables, you can prevent this by setting `<dontWrapJar>`
> to true. The exe acts then as a launcher and starts the application specified in `<jar>` ~~or `<classPath><mainClass>`~~

If the launcher is supposed to launch the jar, does this apply for *non executable* jars as well? IMO it would allow a developer to be lulled into a false sense of security regarding the usability of the created jar in a multi-OS (esp. non-windows) environment.

sschuberth commented:
I'm sorry, but I'm not clear what you mean. First of all, what is an executable JAR in your eyes? I assume a JAR that has a `Main-Class` header in its manifest?

All tests except one do *not* set a `Main-Class`, yet launching the executable works. Thus it indeed looks as if Launch4j's `<mainClass>` is overriding the JAR's `Main-Class` header.

sschuberth commented:
> This is a fix for lazy programmers who deliver non executable jar files.

I think I got now what you mean: You're not happy to set the main class via Launch4j's `<mainClass>` also in the *dontWrapJar* case as you believe the main class should be set in the JAR's manifest, right? While I'd agree to that, I didn't really care about the main class in my change, but about settings the classpath. Theoretically, we should be able to never set Launch4j's `<mainClass>` and always rely on the JAR's manifest, to force "lazy developers" to set the `Main-Class` header.

TheBoegl commented:
IMO an executable jar contains the *Main-Class* and a valid *Class-Path* section in the manifest.
Usually I use the following snippet to create an executable jar in a build script:
```
jar.doFirst {
    manifest {
        def specTitle;
        try {
            specTitle = appName
        } catch (MissingPropertyException ignored) {
            specTitle = project.name
        }
        attributes["Manifest-Version"]="1.0"
        attributes["Created-By"]=vendor
        attributes["Specification-Title"]=specTitle
        attributes["Specification-Version"]=version
        attributes["Specification-Vendor"]=vendor
        attributes["Implementation-Title"]=project.group + '.' + project.name
        attributes["Implementation-Version"]=version
        attributes["Implementation-Vendor"]=vendor
        if (!configurations.compile.isEmpty()) {
            attributes["Class-Path"] = configurations.compile.collect { it.getName() }.join(' ')
        }
        if (project.plugins.hasPlugin('application') && mainClassName != null && !mainClassName.isEmpty()) {
            attributes["Main-Class"]=mainClassName
        }
    }
}
```
(Though this can be written in a groovier way the intentions should be clear, even for programmers new to gradle/groovy.)

The resulting jar could be run easily by double-clicking it - even on a *nix system, if configured properly. Because of the former I was not aware of a real *use case* for providing an option to manually set the classpath or the main class while running the jar directly, but I was wrong. E.g. there might be use cases to create only one jar and more than one executable (see #7) to call different main classes which provide other features...
